backend/src/plugins/camera/camera_service_impl.h

```diff
@@ -121,6 +123,542 @@ class CameraServiceImpl final : public rpc::camera::CameraService::Service {
         return grpc::Status::OK;
     }
 
+    grpc::Status SetMode(grpc::ServerContext * /* context */,
+                         const rpc::camera::SetModeRequest *request,
+                         rpc::camera::SetModeResponse *response) override
+    {
+        if (request != nullptr) {
+            auto camera_result = _camera.set_mode(translateRPCCameraMode(request->camera_mode()));
+
+            if (response != nullptr) {
+                fillResponseWithResult(response, camera_result);
+            }
+        }
+
+        return grpc::Status::OK;
+    }
+
+    static dronecode_sdk::Camera::Mode translateRPCCameraMode(const rpc::camera::CameraMode mode)
+    {
+        switch (mode) {
+            case rpc::camera::CameraMode::PHOTO:
+                return dronecode_sdk::Camera::Mode::PHOTO;
+            case rpc::camera::CameraMode::VIDEO:
+                return dronecode_sdk::Camera::Mode::VIDEO;
+            default:
+                return dronecode_sdk::Camera::Mode::UNKNOWN;
+        }
+    }
+
+    static rpc::camera::CameraMode translateCameraMode(const dronecode_sdk::Camera::Mode mode)
+    {
+        switch (mode) {
+            case dronecode_sdk::Camera::Mode::PHOTO:
+                return rpc::camera::CameraMode::PHOTO;
+            case dronecode_sdk::Camera::Mode::VIDEO:
+                return rpc::camera::CameraMode::VIDEO;
+            default:
+                return rpc::camera::CameraMode::UNKNOWN;
+        }
+    }
+
+    grpc::Status SubscribeMode(grpc::ServerContext * /* context */,
+                               const rpc::camera::SubscribeModeRequest * /* request */,
+                               grpc::ServerWriter<rpc::camera::ModeResponse> *writer) override
+    {
+        std::promise<void> stream_closed_promise;
+        auto stream_closed_future = stream_closed_promise.get_future();
+
+        bool is_finished = false;
```

Comment:
This flag is used to avoid setting the promise twice, right? Why is that happening in the first place?

Reply:
I had the issue in the unit tests, actually. But it's true that there, I am manually triggering the callback, which may never happen from the plugin.

I think this is the first example of a stream that gets disconnected from the client (e.g. the client calls `tryCancel()`). And I need an extra message after cancel has been called (supposedly because of how gRPC works), which was triggering this issue.

I'd say we keep it like this for now, I could investigate further into it later.